adapters/bidmachine/bidmachine.go

```diff
@@ -0,0 +1,221 @@
+package bidmachine
+
+import (
+	"encoding/json"
+	"errors"
+	"fmt"
+	"net/http"
+	"net/url"
+	"path"
+	"strconv"
+	"text/template"
+
+	"github.com/mxmCherry/openrtb"
+	"github.com/prebid/prebid-server/adapters"
+	"github.com/prebid/prebid-server/config"
+	"github.com/prebid/prebid-server/errortypes"
+	"github.com/prebid/prebid-server/macros"
+	"github.com/prebid/prebid-server/openrtb_ext"
+)
+
+type adapter struct {
+	endpoint template.Template
+}
+
+func (a *adapter) MakeRequests(request *openrtb.BidRequest, _ *adapters.ExtraRequestInfo) ([]*adapters.RequestData, []error) {
+	headers := http.Header{}
+	headers.Add("Content-Type", "application/json;charset=utf-8")
+	headers.Add("Accept", "application/json")
+	headers.Add("x-openrtb-version", "2.5")
+
+	impressions := request.Imp
+	result := make([]*adapters.RequestData, 0, len(impressions))
+	errs := make([]error, 0, len(impressions))
+
+	for _, impression := range impressions {
+		if impression.Banner != nil {
+			banner := impression.Banner
+			if banner.Format == nil {
+				errs = append(errs, &errortypes.BadInput{
+					Message: "Impression with id: " + impression.ID + " has following error: Banner format required",
+				})
+				continue
+			}
+			if len(banner.Format) == 0 {
+				errs = append(errs, &errortypes.BadInput{
+					Message: "Impression with id: " + impression.ID + " has following error: Banner format array is empty",
+				})
+				continue
+			}
+		}
+
+		var bidderExt adapters.ExtImpBidder
+		err := json.Unmarshal(impression.Ext, &bidderExt)
+		if err != nil {
+			errs = append(errs, err)
+			continue
+		}
+
+		var impressionExt openrtb_ext.ExtImpBidmachine
+		err = json.Unmarshal(bidderExt.Bidder, &impressionExt)
+		if err != nil {
+			errs = append(errs, err)
+			continue
+		}
+		url, err := a.buildEndpointURL(impressionExt)
+		if err != nil {
+			errs = append(errs, err)
+			continue
+		}
+		if impressionExt.SellerID == "" {
+			errs = append(errs, errors.New("Impression with id: "+impression.ID+" has following error: Bidmachine seller_id is required"))
+			continue
+		}
+		if impressionExt.Path == "" {
+			errs = append(errs, errors.New("Impression with id: "+impression.ID+" has following error: Bidmachine path is required"))
+			continue
+		}
+		if bidderExt.Prebid != nil && bidderExt.Prebid.IsRewardedInventory == 1 {
+			if impression.Banner != nil && !hasRewardedBattr(impression.Banner.BAttr) {
+				bannerCopy := *impression.Banner
+				bannerCopy.BAttr = append(bannerCopy.BAttr, openrtb.CreativeAttribute(16))
```

Comment:
Thank you making the banner copy. I didn't realize at the time, but the backing array for `BAttr` is also shared memory and a potential a race condition. It might be best to add a helper method along the lines of:

```
bannerCopy.BAttr = copyBAttrWithRewardedInventory(bannerCopy.BAttr)
```

```
videoCopy.BAttr = copyBAttrWithRewardedInventory(videoCopy.BAttr)
```

```
func copyBAttrWithRewardedInventory(src []CreativeAttribute) []CreativeAttribute {
  dst = make([]CreativeAttribute, len(src)+1)
  copy(dst, src)
  dst = append(dst, openrtb.CreativeAttribute(16))
  return dst
}
```

Reply:
Fixed, thanks for providing the code!